client consistency proof

macovedj opened the pull request:
Addresses https://github.com/bytecodealliance/registry/issues/82

macovedj commented:
So far this PR just stores checkpoints anytime they're updated so that and old root and new root can be used to hit the consistency endpoint that the warg server offers.  Hoping to finalize and use the response to evaluate a proof soon.

Kylebrown9 commented on crates/client/src/storage/fs.rs:
```diff
@@ -118,6 +121,16 @@ impl PackageStorage for FileSystemPackageStorage {
     }
 
+    async fn load_checkpoint(&self) -> Result<Option<SerdeEnvelope<MapCheckpoint>>> {
+        let checkpoint = load(&self.package_path("checkpoint")).await?;
```
I don't think this is the correct place to put the checkpoint. This is the path for a package named `checkpoint`. 

This should probably be a new file path with its own function & hard-coded name like `pending_publish_path`.

macovedj commented on crates/client/src/storage/fs.rs:
looks like this is the only one right now

Kylebrown9 commented:
The proof/API side of this looks good, but I'm not sure about the storage/CLI part.

@peterhuene can you take a look at it?

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -85,6 +85,15 @@ pub struct Config {
+    /// If `None`, the default of `$CACHE_DIR/warg/checkpoint` is used, where
+    /// `$CACHE_DIR` is the platform-specific cache directory.
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    pub checkpoint_path: Option<PathBuf>,
```
I don't think this is correct. The checkpoint should be stored per-registry, and this appears to be a single checkpoint for the client regardless of registry.

Perhaps `packages_dir` should become `registries_dir` (`$CACHE_DIR/warg/registries`), which underneath it has:

* `packages` - a directory containing the package logs
* `checkpoint` - a file containing the last known registry checkpoint
* `operator.log` - the registry's operator log (for the future, don't need to implement this now as we're not doing anything with the operator log currently).

macovedj replied:
Ah this resonates.  I was actually looking into this [client operator issue](https://github.com/bytecodealliance/registry/issues/83) a bit this morning, and was thinking it would require something like what you're suggesting.  I'll try and have some changes up soon.

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -228,10 +239,24 @@ impl Config {
     }
 
+    /// Gets the path to the file where the previously fetched checkpoint is stored
+    pub fn checkpoint_path(&self) -> Result<PathBuf> {
```
As a consequence of the above, I'd expect this method to be removed.

peterhuene commented on crates/client/src/lib.rs:
```diff
@@ -33,20 +34,22 @@ pub mod storage;
 }
 
-impl<P: PackageStorage, C: ContentStorage> Client<P, C> {
+impl<P: PackageStorage, C: ContentStorage, T: CheckpointStorage> Client<P, C, T> {
```
I think we should probably rename `PackageStorage` to `RegistryStorage` and incorporate the checkpoint into it; eventually we'll need to add in operator log reading/writing into it and I'd rather not have more traits.

The reason there's two traits here is that content storage is across all registries, whereas the rest of the data is per-registry.

peterhuene reviewed (commented):
Thanks so much for taking these changes on, @macovedj!

I have some feedback comments to address. Most of the feedback stems from the existing design needed to be refactored to accommodate more information about this and future needs.

peterhuene commented on crates/client/src/api.rs:
```diff
@@ -219,9 +225,48 @@ impl Client {
-        todo!()
+        let old = old_root.clone();
+        let new = new_root.clone();
+        let request = ConsistencyRequest { old_root, new_root };
```
Side note: the request types in the API crate should borrow their data so we don't have these unnecessary clones.

I'll file an issue.

peterhuene replied:
Filed as https://github.com/bytecodealliance/registry/issues/104

macovedj commented:
@peterhuene I think that I addressed most of the comments made.  For the time being, I didn't see an alternative to hardcoding the registries that we expect to interact with and want to be a part of local storage.  I'm thinking that the eventual solution is to have the server responses include something like a `registry_name` that would be used to determine which subfolder package information would get stored in, but wasn't sure whether or not that should get scoped into this PR.

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -60,6 +60,15 @@ fn normalize_path(path: &Path) -> PathBuf {
+pub struct StoragePaths {
+    /// url path
+    pub url: Url,
+    /// registries path
```
```suggestion
    /// The path to the registry storage directory.
```

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -60,6 +60,15 @@ fn normalize_path(path: &Path) -> PathBuf {
+    pub url: Url,
+    /// registries path
+    pub registries_dir: PathBuf,
+    /// content path
```
```suggestion
    /// The path to the content storage directory.
```

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -68,14 +77,17 @@ pub struct Config {
     pub default_url: Option<String>,
 
+    /// List of registries
+    pub registries: Vec<String>,
```
I can't seem to determine what this is used for. Could you describe the intention behind this field?

Kylebrown9 replied:
Bump @macovedj

macovedj replied:
Peter's comment helped me realize it wasn't necessary.  Went ahead and deleted this field.

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -68,14 +77,17 @@ pub struct Config {
+    /// List of registries
+    pub registries: Vec<String>,
+
     /// The path to the directory where per-registry packages are stored.
```
```suggestion
    /// The path to the top-level directory where per-registry information is stored.
```

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -68,14 +77,17 @@ pub struct Config {
     ///
     /// This path is expected to be relative to the configuration file.
     ///
     /// If `None`, the default of `$CACHE_DIR/warg/packages` is used, where
```
```suggestion
    /// If `None`, the default of `$CACHE_DIR/warg/registries` is used, where
```

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -201,8 +214,8 @@ impl Config {
     /// Gets the path to the directory where per-registry packages are stored.
-    pub fn packages_dir(&self) -> Result<PathBuf> {
-        self.packages_dir
+    pub fn registries_dir(&self) -> Result<PathBuf> {
```
Can we update the default in here from `warg/packages` to `warg/registries`?

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -231,15 +244,19 @@ impl Config {
 
         let host = url.host().unwrap().to_string().to_ascii_lowercase();
-        let packages_dir = self.packages_dir()?.join(host);
+        let registries_dir = self.registries_dir()?.join(host);
```
As a note, this is the line where we hand out the per-registry storage location based on the host of the registry we're communicating with.

peterhuene commented on crates/client/src/lib.rs:
```diff
@@ -71,12 +72,12 @@ impl<P: PackageStorage, C: ContentStorage> Client<P, C> {
         let info = self
             .packages
-            .load_publish()
+            .load_publish("dogfood")
```
The `RegistryStorage` is already per-registry, so no need for a hardcoded `"dogfood"` here and below (see my comments on the `RegistryStorage` trait).

peterhuene commented on crates/client/src/lib.rs:
```diff
@@ -39,7 +40,7 @@ pub struct Client<P, C> {
 }
 
-impl<P: PackageStorage, C: ContentStorage> Client<P, C> {
+impl<P: RegistryStorage, C: ContentStorage> Client<P, C> {
```
There's a few places here where comments refer to "package" storage that should be updated to "registry storage".

peterhuene replied:
Also, should rename the `packages` method (perhaps simply to `registry`?).

peterhuene commented on crates/client/src/storage.rs:
```diff
@@ -24,27 +24,38 @@ pub use fs::*;
-pub trait PackageStorage: Send + Sync {
+pub trait RegistryStorage: Send + Sync {
+    ///  Loads most recent checkpoint
+    async fn load_checkpoint(&self, registry: &str)
```
The `RegistryStorage` is already specific to a particular registry; we should revert any changes to the existing methods from the previous `PackageStorage` trait.

We should also update the comments on the trait to something like:

```
/// Trait for registry storage implementations.
///
/// Stores information such as package/operator logs and checkpoints
/// on a per-registry basis.
///
/// Registry storage data must be synchronized if shared between
/// multiple threads and processes.
```

macovedj commented:
On second thought, I think I can get around the hardcoding for the time being, but still have open questions around whether registry identifiers need to be included in endpoint responses.  Should be able to push something to remove the hardcoding tomorrow.

peterhuene reviewed (commented):
The updated changes look like a step in the right direction.

I think we can remove the hardcoded "dogfood" references (see individual review comments).

macovedj commented:
Ok I think it's ready for review again.
One final thing with respect to this comment:

> I don't think this is correct. The checkpoint should be stored per-registry, and this appears to be a single checkpoint for the client regardless of registry.
> 
> Perhaps `packages_dir` should become `registries_dir` (`$CACHE_DIR/warg/registries`), which underneath it has a sub-directory per registry (e.g. `registry.bytecodealliance.org`) containing:
> 
> * `packages` - a directory containing the package logs for the registry
> * `checkpoint` - a file containing the last known registry checkpoint
> * `operator.log` - the registry's operator log (for the future, don't need to implement this now as we're not doing anything with the operator log currently).

The current changes in this PR rename `packages` to `registries`, and the only difference in the structure of its contents is that in a host subfolder, there's now a file called `checkpoint` which will store the content from the most recent checkpoint update for that host/registry server.  This didn't require any changes to any structs, just added a `.join("checkpoint")` in the new methods.  Since the methods for `operator.log` don't exist yet, and since a new struct wasn't necessary, there wasn't a need for any operator-related changes to appear in this PR, but I'm guessing that we'll see changes in the future PR that handles operator verification on the client.

I considered adding some reference to a a checkpoint path to to the `StoragePaths` struct, but it seemed like it would complicate things more than it would help.

peterhuene commented on crates/api/src/proof.rs:
```diff
@@ -91,6 +91,15 @@ pub enum ProofError {
         id: LogId,
     },
+
+    /// Failed to prove consistency of a package.
```
```suggestion
    /// Failed to prove consistency of the registry between two checkpoints.
```

peterhuene commented on crates/client/src/config.rs:
```diff
@@ -60,6 +60,15 @@ fn normalize_path(path: &Path) -> PathBuf {
+    pub registries_dir: PathBuf,
+    /// The path to the content storage directory.
+    pub content_dir: PathBuf,
+}
```
Nit: 

```suggestion
}

```

peterhuene commented on crates/client/src/storage.rs:
```diff
@@ -19,12 +19,21 @@ use warg_protocol::{
+    ///  Loads most recent checkpoint
+    async fn load_checkpoint(&self) -> Result<Option<SerdeEnvelope<MapCheckpoint>>>;
+
+    ///  Stores most recent checkpoint
```
Nit:

```suggestion
    /// Stores most recent checkpoint
```

peterhuene commented on crates/client/src/storage.rs:
```diff
@@ -19,12 +19,21 @@ use warg_protocol::{
 #[async_trait]
-pub trait PackageStorage: Send + Sync {
+pub trait RegistryStorage: Send + Sync {
+    ///  Loads most recent checkpoint
```
Nit:

```suggestion
    /// Loads most recent checkpoint
```

peterhuene commented on crates/client/src/storage/fs.rs:
```diff
@@ -76,7 +79,16 @@ impl FileSystemPackageStorage {
+    }
+
+    async fn store_checkpoint(&self, checkpoint: SerdeEnvelope<MapCheckpoint>) -> Result<()> {
+        store(&self.base_dir.join("checkpoint"), checkpoint).await;
+        Ok(())
```
This appears to be an unused result:

```suggestion
        store(&self.base_dir.join("checkpoint"), checkpoint).await
```

peterhuene commented on crates/client/src/storage/fs.rs:
```diff
@@ -76,7 +79,16 @@ impl FileSystemPackageStorage {
-impl PackageStorage for FileSystemPackageStorage {
+impl RegistryStorage for FileSystemPackageStorage {
+    async fn load_checkpoint(&self) -> Result<Option<SerdeEnvelope<MapCheckpoint>>> {
+        let contents = load(&self.base_dir.join("checkpoint")).await;
+        contents
```
Nit (as it's not really the contents, just the result):

```suggestion
        load(&self.base_dir.join("checkpoint")).await
```

peterhuene commented on src/commands/config.rs:
```diff
@@ -12,7 +12,7 @@ pub struct ConfigCommand {
     /// The path to the packages directory to use.
     #[clap(long, value_name = "PACKAGES")]
-    pub packages_dir: Option<PathBuf>,
+    pub registries_dir: Option<PathBuf>,
```
Need to update the comment on this field and the value to `REGISTRIES`.

peterhuene replied:
I think the `value_name` needs updating still and that's the last of it 👍

peterhuene reviewed (commented):
Looks great!

Thanks for putting up with all my comments.

The remaining is almost entirely nits at this point, except for one unused result warning that needs fixing.

macovedj commented:
> Looks great!
> 
> Thanks for putting up with all my comments.
> 
> The remaining is almost entirely nits at this point, except for one unused result warning that needs fixing.

Appreciate it!  Thanks for putting up with my missing some of the comments that needed changing and such.  I'll be more vigilant on the next PR to avoid us having so much to cover.

peterhuene commented on crates/client/src/storage.rs:
```diff
@@ -19,12 +19,21 @@ use warg_protocol::{
+    async fn load_checkpoint(&self) -> Result<Option<SerdeEnvelope<MapCheckpoint>>>;
+
+    /// Stores most recent checkpoint
+    async fn store_checkpoint(&self, checkpoint: SerdeEnvelope<MapCheckpoint>);
```
I'd expect `store_checkpoint` to still return `Result<()>` as it is a fallible operation.

macovedj replied:
Brought back the return type

peterhuene reviewed (approved):
Looks great 👍

Thanks for implementing this!
